### DISCUSSION TITLE
event.keyCode is deprecated - should we use event.key in `$tw.keyboardManager`?

### DISCUSSION THREAD

**Discussion by BurningTreeC** [TiddlyWiki/TiddlyWiki5#8753] [General]
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode

There's a big red box stating that `event.keyCode` is deprecated in favor of `event.key`

For the sake of future compatibility and simplicity, shouldn't we use `event.key` in `$tw.keyboardManager`?

**Comment by pmario** [comment]
Can you link to the file(s) at GH where where you think we should change it. 

I did a global search for `event.keycode` -- It is used 15 times in 9 files in several plugins. I am pretty sure, that this element will stay for quite some time. I think it is not urgent.

I found at the tw modules, where we can change them

- https://github.com/TiddlyWiki/TiddlyWiki5/blob/5ea407738558541405f335f736c74779e49320eb/core/modules/keyboard.js#L276
- https://github.com/TiddlyWiki/TiddlyWiki5/blob/5ea407738558541405f335f736c74779e49320eb/core/modules/widgets/edit-shortcut.js#L105-L112

**Reply by BurningTreeC** [reply]
Hi @pmario - these are exactly the places where they should be changed.
Then also the long object `namedKeys` wouldn't be needed anymore - we wouldn't need to convert keyCodes to named keys and vice versa

**Reply by BurningTreeC** [reply]
I have already a working prototype

**Reply by pmario** [reply]
If it works already, you should create a PR. IMO it's worth it. 

But as I wrote, we do have several 3rd party plugins, that use libraries which also use `event.keyCode` ... So I think browser vendors will keep the function for quite some time.

**Comment by saqimtiaz** [comment]
> For the sake of future compatibility and simplicity, shouldn't we use `event.key` in `$tw.keyboardManager`?

I don't believe we can do so until we officially stop supporting Internet Explorer. As I recall, event.key is not properly supported in IE.

**Reply by BurningTreeC** [reply]
> > For the sake of future compatibility and simplicity, shouldn't we use `event.key` in `$tw.keyboardManager`?
> 
> I don't believe we can do so until we officially stop supporting Internet Explorer. As I recall, event.key is not properly supported in IE.

That's too bad. It seems to be partially supported but broken: https://caniuse.com/?search=event.key

I would need this for the CodeMirror 6 plugin. I would like to add Keyboard Shortcuts through TiddlyWiki's UI.
CodeMirror internally doesn't use shortcuts like ctrl-Comma but uses ctrl-,

**Reply by BurningTreeC** [reply]
> > > For the sake of future compatibility and simplicity, shouldn't we use `event.key` in `$tw.keyboardManager`?
> > 
> > 
> > I don't believe we can do so until we officially stop supporting Internet Explorer. As I recall, event.key is not properly supported in IE.
> 
> That's too bad. It seems to be partially supported but broken: https://caniuse.com/?search=event.key
> 
> I would need this for the CodeMirror 6 plugin. I would like to add Keyboard Shortcuts through TiddlyWiki's UI. CodeMirror internally doesn't use shortcuts like ctrl-Comma but uses ctrl-,

I will then maybe need a conversion table

**Comment by pmario** [comment]
@saqimtiaz ... IMO this one should be added to https://github.com/orgs/TiddlyWiki/projects/4

**Reply by saqimtiaz** [reply]
@pmario It looks like only PRs and Issues can be added to project plans.

**Reply by pmario** [reply]
ahhh. OK. Can you convert it into an issue?

**Reply by pmario** [reply]
@saqimtiaz -- I saw you did create a completely new issue.

I thought there would be a button here in the discussion UI, to convert the discussion into an issue?

**Reply by saqimtiaz** [reply]
@pmario that is what converting the discussion into an issue does, it creates an entirely new issue linked to the discussion.

**Comment by BurningTreeC** [comment]
Hi all, I'll create a PR now that addresses this discussion

**Comment by saqimtiaz** [comment]
Please continue the discussion at https://github.com/TiddlyWiki/TiddlyWiki5/issues/9140
